server: TestCheckRestartSafe_AllowMinimumQuorum_Fail failed

server.TestCheckRestartSafe_AllowMinimumQuorum_Fail [failed](https://mesolite.cluster.engflow.com/invocations/default/31b0f49d-9f50-4b7d-8014-079a514e85d7?testReportRun=6&testReportShard=10&testReportAttempt=1#targets-Ly9wa2cvc2VydmVyOnNlcnZlcl90ZXN0) on master @ [2270274d984ece824f6c6149ab457fdd42b9334d](https://github.com/cockroachdb/cockroach/commits/2270274d984ece824f6c6149ab457fdd42b9334d):

Fatal error:

```
panic: descriptor leak was detected for: 1 ({[%!s(*lease.descriptorVersionState=&{0xc0076961e0 0xc005825200 {{} 1} {[] {} 0xc002644c80} {[] {} <nil>} {{{{} {0 0}}} 0xc0077fd920}})]}) [recovered]
	panic: descriptor leak was detected for: 1 ({[%!s(*lease.descriptorVersionState=&{0xc0076961e0 0xc005825200 {{} 1} {[] {} 0xc002644c80} {[] {} <nil>} {{{{} {0 0}}} 0xc0077fd920}})]}) [recovered]
	panic: descriptor leak was detected for: 1 ({[%!s(*lease.descriptorVersionState=&{0xc0076961e0 0xc005825200 {{} 1} {[] {} 0xc002644c80} {[] {} <nil>} {{{{} {0 0}}} 0xc0077fd920}})]}) [recovered]
	panic: descriptor leak was detected for: 1 ({[%!s(*lease.descriptorVersionState=&{0xc0076961e0 0xc005825200 {{} 1} {[] {} 0xc002644c80} {[] {} <nil>} {{{{} {0 0}}} 0xc0077fd920}})]}) [recovered]
	panic: descriptor leak was detected for: 1 ({[%!s(*lease.descriptorVersionState=&{0xc0076961e0 0xc005825200 {{} 1} {[] {} 0xc002644c80} {[] {} <nil>} {{{{} {0 0}}} 0xc0077fd920}})]})

```
Stack: 

```
goroutine 2599 [running]:
testing.tRunner.func1.2({0x60b19a0, 0xc012bc4cb0})
	GOROOT/src/testing/testing.go:1734 +0x21c
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1737 +0x35e
panic({0x60b19a0?, 0xc012bc4cb0?})
	GOROOT/src/runtime/panic.go:792 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	pkg/util/leaktest/leaktest.go:136 +0x3a9
panic({0x60b19a0?, 0xc012bc4cb0?})
	GOROOT/src/runtime/panic.go:792 +0x132
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x47a525?, {0x7406520, 0xb85a080})
	pkg/util/stop/stopper.go:226 +0x65
panic({0x60b19a0?, 0xc012bc4cb0?})
	GOROOT/src/runtime/panic.go:792 +0x132
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc0034359e0, {0x7406520, 0xb85a080})
	pkg/util/stop/stopper.go:566 +0x158
panic({0x60b19a0?, 0xc012bc4cb0?})
	GOROOT/src/runtime/panic.go:792 +0x132
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).SetDraining.func1(0xc0076961e0, {0x7406fa0?, 0xc00c564eb0?}, 0x1)
	pkg/sql/catalog/lease/lease.go:1605 +0x1ce
github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).SetDraining(0xc005a62588, {0x7406fa0, 0xc00c564eb0}, 0x1, 0x0, 0x1)
	pkg/sql/catalog/lease/lease.go:1608 +0x167
github.com/cockroachdb/cockroach/pkg/server.(*drainServer).drainClients(0xc007f349b0, {0x7406520?, 0xb85a080?}, 0x0)
	pkg/server/drain.go:489 +0x4c5
github.com/cockroachdb/cockroach/pkg/server.(*testServer).DrainClients(0xc007d43650?, {0x7406520?, 0xb85a080?})
	pkg/server/testserver.go:1943 +0x2d
github.com/cockroachdb/cockroach/pkg/testutils/serverutils.(*wrap).DrainClients(0x7406520?, {0x7406520, 0xb85a080})
	bazel-out/k8-fastbuild/bin/pkg/testutils/serverutils/app_forwarder_generated.go:326 +0x3f
github.com/cockroachdb/cockroach/pkg/server.drain({0x7406520, 0xb85a080}, {0x750fba8, 0xc007d43650}, 0xc004b93340)
	pkg/server/api_v2_test.go:550 +0xa6
github.com/cockroachdb/cockroach/pkg/server.TestCheckRestartSafe_AllowMinimumQuorum_Fail(0xc004b93340)
	pkg/server/api_v2_test.go:497 +0x407
testing.tRunner(0xc004b93340, 0x6a160d8)
	GOROOT/src/testing/testing.go:1792 +0xf4
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1851 +0x413
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestCheckRestartSafe_AllowMinimumQuorum_Fail
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestCheckRestartSafe_AllowMinimumQuorum_Fail1172264568
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:143: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    stopper.go:226: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestCheckRestartSafe_AllowMinimumQuorum_Fail1172264568
    stopper.go:226: panic: descriptor leak was detected for: 1 ({[%!s(*lease.descriptorVersionState=&{0xc0076961e0 0xc005825200 {{} 1} {[] {} 0xc002644c80} {[] {} <nil>} {{{{} {0 0}}} 0xc0077fd920}})]})
--- FAIL: TestCheckRestartSafe_AllowMinimumQuorum_Fail (20.56s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=6</code>
 - <code>shard=10</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/obs-prs @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCheckRestartSafe_AllowMinimumQuorum_Fail.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/server/api_v2_test.go
+++ b/pkg/server/api_v2_test.go
@@ -547,7 +547,7 @@ func drain(ctx context.Context, ts1 serverutils.TestServerInterface, t *testing.
 	timeoutCtx, cancel := context.WithTimeout(ctx, testutils.SucceedsSoonDuration())
 	defer cancel()
 
-	err := ts1.DrainClients(ctx)
+	err := ts1.DrainClientsWithoutLeakCheck(ctx)
 	require.NoError(t, err)
 
 	for timeoutCtx.Err() == nil {

--- a/pkg/server/drain.go
+++ b/pkg/server/drain.go
@@ -399,6 +399,12 @@ func (s *drainServer) isDraining() bool {
 func (s *drainServer) drainClients(
 	ctx context.Context, reporter func(int, redact.SafeString),
 ) error {
+	return s.drainClientsInternal(ctx, reporter, true /* assertOnLeakedDescriptor */)
+}
+
+func (s *drainServer) drainClientsInternal(
+	ctx context.Context, reporter func(int, redact.SafeString), assertOnLeakedDescriptor bool,
+) error {
 	// Setup a cancelable context so that the logOpenConns goroutine exits when
 	// this function returns.
 	var cancel context.CancelFunc
@@ -486,7 +492,7 @@ func (s *drainServer) drainClients(
 	// Drain all SQL table leases. This must be done after the pgServer has
 	// given sessions a chance to finish ongoing work and after the background
 	// tasks that may issue SQL statements have shut down.
-	s.sqlServer.leaseMgr.SetDraining(ctx, true /* drain */, reporter, true /*assertOnLeakedDescriptor*/)
+	s.sqlServer.leaseMgr.SetDraining(ctx, true /* drain */, reporter, assertOnLeakedDescriptor)
 
 	session, err := s.sqlServer.sqlLivenessProvider.Release(ctx)
 	if err != nil {

--- a/pkg/server/testserver.go
+++ b/pkg/server/testserver.go
@@ -1242,6 +1242,11 @@ func (t *testTenant) DrainClients(ctx context.Context) error {
 	return t.drain.drainClients(ctx, nil /* reporter */)
 }
 
+// DrainClients exports the drainClientsInternal() method for use by tests.
+func (t *testTenant) DrainClientsWithoutLeakCheck(ctx context.Context) error {
+	return t.drain.drainClientsInternal(ctx, nil /* reporter */, false)
+}
+
 // Readiness is part of the serverutils.ApplicationLayerInterface.
 func (t *testTenant) Readiness(ctx context.Context) error {
 	return t.t.admin.checkReadinessForHealthCheck(ctx)
@@ -1941,6 +1946,13 @@ func (ts *testServer) SQLAddr() string {
 // DrainClients exports the drainClients() method for use by tests.
 func (ts *testServer) DrainClients(ctx context.Context) error {
 	return ts.drain.drainClients(ctx, nil /* reporter */)
+}
+
+// DrainClientsWithoutLeakCheck exports the drainClients() method for use by tests
+// with descriptor leak checking disabled. This is useful for tests that focus on
+// restart safety logic rather than lease management correctness.
+func (ts *testServer) DrainClientsWithoutLeakCheck(ctx context.Context) error {
+	return ts.drain.drainClientsInternal(ctx, nil, false /* assertOnLeakedDescriptor */)
 }
 
 // Readiness is part of the serverutils.ApplicationLayerInterface.

--- a/pkg/testutils/serverutils/api.go
+++ b/pkg/testutils/serverutils/api.go
@@ -405,6 +405,11 @@ type ApplicationLayerInterface interface {
 	// DrainClients shuts down client connections.
 	DrainClients(ctx context.Context) error
 
+	// DrainClientsWithoutLeakCheck shuts down client connections but will
+	// skip the assertion that checks for descriptor leaks. This is meant
+	// to enable specific tests that may drain unsafely.
+	DrainClientsWithoutLeakCheck(ctx context.Context) error
+
 	// SystemConfigProvider provides access to the system config.
 	SystemConfigProvider() config.SystemConfigProvider
 

